flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSourceSinkFactory.java

```diff
@@ -0,0 +1,228 @@
+	public static final String CONNECTOR_SCAN_LOWER_BOUND = "connector.scan.lower-bound";
+	public static final String CONNECTOR_SCAN_UPPER_BOUND = "connector.scan.upper-bound";
+	public static final String CONNECTOR_SCAN_NUM_PARTITIONS = "connector.scan.num-partitions";
+
+	public static final String CONNECTOR_LOOKUP_CACHE_MAX_ENTRIES = "connector.lookup.cache.max-entries";
+	public static final String CONNECTOR_LOOKUP_CACHE_TTL = "connector.lookup.cache.ttl";
+	public static final String CONNECTOR_LOOKUP_MAX_RETRIES = "connector.lookup.max-retries";
+
+	public static final String CONNECTOR_WRITE_FLUSH_MAX_ENTRIES = "connector.write.flush.max-size";
+	public static final String CONNECTOR_WRITE_FLUSH_INTERVAL = "connector.write.flush.interval";
+	public static final String CONNECTOR_WRITE_MAX_RETRIES = "connector.write.max-retries";
+
+	@Override
+	public Map<String, String> requiredContext() {
+		Map<String, String> context = new HashMap<>();
+		context.put(CONNECTOR_TYPE, "jdbc"); // jdbc
+		context.put(CONNECTOR_PROPERTY_VERSION, "1"); // backwards compatibility
+		return context;
+	}
+
+	@Override
+	public List<String> supportedProperties() {
+		List<String> properties = new ArrayList<>();
+
+		// common options
+		properties.add(CONNECTOR_DRIVER);
+		properties.add(CONNECTOR_URL);
+		properties.add(CONNECTOR_DBTABLE);
+		properties.add(CONNECTOR_USER);
+		properties.add(CONNECTOR_PASSWORD);
+
+		// scan partition options
+		properties.add(CONNECTOR_SCAN_PARTITION_COLUMN);
+		properties.add(CONNECTOR_SCAN_NUM_PARTITIONS);
+		properties.add(CONNECTOR_SCAN_LOWER_BOUND);
+		properties.add(CONNECTOR_SCAN_UPPER_BOUND);
+
+		// lookup options
+		properties.add(CONNECTOR_LOOKUP_CACHE_MAX_ENTRIES);
+		properties.add(CONNECTOR_LOOKUP_CACHE_TTL);
+		properties.add(CONNECTOR_LOOKUP_MAX_RETRIES);
+
+		// sink options
+		properties.add(CONNECTOR_WRITE_FLUSH_MAX_ENTRIES);
+		properties.add(CONNECTOR_WRITE_FLUSH_INTERVAL);
+		properties.add(CONNECTOR_WRITE_MAX_RETRIES);
+
+		// schema
+		properties.add(SCHEMA + ".#." + SCHEMA_TYPE);
+		properties.add(SCHEMA + ".#." + SCHEMA_NAME);
+
+		return properties;
+	}
+
+	@Override
+	public StreamTableSource<Row> createStreamTableSource(Map<String, String> properties) {
+		final DescriptorProperties descriptorProperties = getValidatedProperties(properties);
+
+		final JDBCOptions options = getJDBCOptions(descriptorProperties);
+
+		final Optional<String> partitionColumnName =
+			descriptorProperties.getOptionalString(CONNECTOR_SCAN_PARTITION_COLUMN);
+		final Optional<Long> partitionLower = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_LOWER_BOUND);
+		final Optional<Long> partitionUpper = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_UPPER_BOUND);
+		final Optional<Long> numPartitions = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_NUM_PARTITIONS);
+		final JDBCScanOptions scanOptions;
+		if (partitionColumnName.isPresent()) {
```

Comment:
Use `Optional.ifPresent`? That will reduce a lot of lines.

Reply:
+1. You can refer this: https://github.com/apache/flink/pull/9075/files#diff-e63010407ba50e1594b5d72607b15931